src/main/java/pl/edu/icm/pl/mxrdr/extension/workflow/XdsSupplyMissingValuesStep.java

```diff
@@ -0,0 +1,137 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class XdsSupplyMissingValuesStep extends FilesystemAccessingWorkflowStep {
+
+    static final String STEP_ID = "xds-supply-missing-values";
+
+    private static final Set<String> SOURCE_FIELDS_SET = Initializer.initializeSourceFieldsSet();
+    private static final Map<String, PatternAndReplacement> PATTERNS = Initializer.initializePatternMap();
+
+    private DatasetVersionServiceBean versionService;
+
+    public XdsSupplyMissingValuesStep(Map<String, String> inputParams, DatasetVersionServiceBean versionService) {
+        super(inputParams);
+        this.versionService = versionService;
+    }
+
+    @Override
+    protected WorkflowStepResult.Source runInternal(WorkflowExecutionContext workflowExecutionContext, Path path) throws Exception {
+        List<PatternAndReplacement> actions = prepareReplacementActions(workflowExecutionContext);
+        FileContentReplacer.of(XdsAdjustResultStep.XDS_INP, path)
+                .add(actions)
+                .replace();
+        return Success::new;
+    }
+
+    @Override
+    public WorkflowStepResult resume(WorkflowExecutionContext workflowExecutionContext, Map<String, String> map, String s) {
+        throw new UnsupportedOperationException("This step does not pause");
+    }
+
+    @Override
+    public void rollback(WorkflowExecutionContext workflowExecutionContext, Failure failure) { }
+
+    // -------------------- PRIVATE --------------------
+
+    private List<PatternAndReplacement> prepareReplacementActions(WorkflowExecutionContext workflowExecutionContext) {
+        Map<String, String> replacementValues = readReplacementValues(workflowExecutionContext);
+        return PATTERNS.keySet().stream()
+                .filter(k -> replacementValues.get(k) != null)
+                .map(k -> {
+                    PatternAndReplacement patterns = PATTERNS.get(k);
+                    return FileContentReplacer.PatternAndReplacement.of(patterns.pattern,
+                            String.format(patterns.replacement, replacementValues.get(k)));
+                })
+                .collect(Collectors.toList());
+    }
+
+    private Map<String, String> readReplacementValues(WorkflowExecutionContext executionContext) {
+        return findVersion(executionContext)
+                .map(DatasetVersion::getDatasetFields)
+                .orElse(Collections.emptyList()).stream()
+                .filter(f -> MxrdrMetadataField.DATA_COLLECTION.getValue()
+                        .equals(extractFieldName(f)))
+                .findFirst()
+                .map(DatasetField::getDatasetFieldsChildren)
+                .orElse(Collections.emptyList()).stream()
+                .filter(f -> SOURCE_FIELDS_SET.contains(extractFieldName(f)))
+                .collect(HashMap::new,
+                        (m, f) -> m.put(extractFieldName(f), f.getValue()),
+                        HashMap::putAll);
```

Comment:
As far as I remember `Collectors.toMap(..)` has a problem with `null` values – it will throw an exception in such case, so I prefer this way, though less elegant indeed.

Reply:
It has a problem with duplicated keys for sure - you can't have those, but that's not an issue here. Even if it had a problem with `null` values, why would you want to replace anything with `null`s? Shouldn't those be filtered out?